presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java

```diff
@@ -528,6 +529,27 @@ public void dropTable(Session session, TableHandle tableHandle)
     }
 
     @Override
+    public void beginQuery(Session session, Collection<TableHandle> tableHandles)
+    {
+        Collection<TableHandle> distinctHandles = filterDuplicateCatalogs(tableHandles); // don't trigger ConnectorMetadata.beginQuery more than once when >1 tables of the same catalog are involved in the query
```

Comment:
Also, would it work to return a set of connector ids here rather than doing the TreeSet logic?